### PR TITLE
Use a separate client process

### PR DESCRIPTION
**Pull request by sorentwo**
This PR changes the how the client works, replacing the use of tasks for a supervised client. It ended up being larger than intended because the changes were difficult to isolate—on the plus side, it covers at least three open issues.

* Drops HTTPoison in favor of directly using Hackney, which gives us access to a connection pool
* Drops Meck and stubbing in favor of a local cowboy server (honeybadger-io/honeybadger-elixir#7)
* Changes `notify` from a macro to a function
* Stops spawning new tasks for every error, instead relying on async handling in the client (honeybadger-io/honeybadger-elixir#88)
* Increases the logging around client activity (honeybadger-io/honeybadger-elixir#20)
* Starts a supervision tree with the client as a child

It is based off of honeybadger-io/honeybadger-elixir#89, largely to prevent merge conflicts and simplify the changes.

### CODE REVIEW

**Comment by joshuap** [comment]
@sorentwo I can't wait to go through this, it sounds great. I'm planning to dig in later this week. Any ideas on the build failures? I can pull this down and run the suite locally--I assume they are  related to the changes around dropping Meck. Assuming the build is passing locally I can probably fix Travis.

By the way, if anyone else feels like giving this a code review, feel free.

**Comment by sorentwo** [comment]
@joshuap The build is passing again. I forgot that `left not in right` wasn't [added until recently[(https://github.com/elixir-lang/elixir/pull/5620), so the older versions were getting syntax errors.

Everything is fixed up now and this has been rebased off master, mostly to get the additional builds from the matrix update.

File: lib/honeybadger/client.ex [hunk 1]
```diff
@@ -1,57 +1,105 @@
...
+  def handle_cast({:notice, notice}, %{enabled: true} = state) do
+    payload = Poison.encode!(notice)
+
+    opts =
+      state
+      |> Map.take([:proxy, :proxy_auth])
+      |> Enum.into(Keyword.new)
+      |> Keyword.put(:pool, __MODULE__)
+
+    case :hackney.post(state.url, state.headers, payload, opts) do
```
[the remark this answers is not in the document]
Comment (sorentwo) [reply lib/honeybadger/client.ex]: It was due to the use of `Application.get_all_env`, which didn't make any use of `Honeybadger.get_env/1`. I've added a new function, `Honeybadger.get_all_env/0` that takes care of this automatically—and is tested.

File: lib/honeybadger/client.ex [hunk 2]
```diff
@@ -1,57 +1,105 @@
...
+      {:ok, code, _headers, body} when code >= 200 and code <= 399 ->
+        Logger.debug("[Honeybadger] API success: #{inspect(body)}")
```
[the remark this answers is not in the document]
Comment (sorentwo) [reply lib/honeybadger/client.ex]: Great catch! That's much better. I've updated the decoding and added some assertions about what gets logged. 👍

**Comment by sorentwo** [comment]
@joshuap All of your feedback should have been addressed. Thanks for looking it over.

File: lib/honeybadger.ex [hunk 3]
```diff
@@ -132,73 +130,87 @@ defmodule Honeybadger do
...
 
-  def context(keyword_or_map) do
-    Process.put(@context, Map.merge(context(), Enum.into(keyword_or_map, %{})))
-    context()
-  end
+  # Helpers
 
   defp default_config do
-     [api_key: System.get_env("HONEYBADGER_API_KEY"),
+     [api_key: {:system, "HONEYBADGER_API_KEY"},
```
Comment (joshuap) [inline lib/honeybadger.ex]: I think this is the source of the config issue. `{:system, "HONEYBADGER_API_KEY"}` and `{:system, "MIX_ENV"}` should be environment variables, but they are being interpreted literally.

Comment (sorentwo) [reply lib/honeybadger.ex]: Of course! The `get_all_env/0` function is being called before the merge, which is backward. I'll rework that now.

**Comment by sorentwo** [comment]
@joshuap Alright, shuffled the config a bit and refined how the defaults are merged or preserved. It is cleaner now, and I've verified that none of the `{:system, _}` tuples make it through to the client:

```elixir
Honeybadger.Client.new([])
%Honeybadger.Client{enabled: false,
 headers: [{"X-API-Key", "asdfghjkl"}, {"Accept", "application/json"},
  {"Content-Type", "application/json"}, {"User-Agent", "Honeybadger Elixir"}],
 proxy: nil, proxy_auth: {nil, nil},
 url: "https://api.honeybadger.io/v1/notices"}
```

**Comment by joshuap** [comment]
@sorentwo thanks, this looks good to me! 👍 I'm going to merge now and release a beta.

**Comment by sorentwo** [comment]
@joshuap Awesome! Thanks for all the feedback and troubleshooting.

**Pull request merged**